File: 2019_07_SocketIO/Alick/SocketIo_done/tests/TestConnection.ts
import {assert, expect} from "chai";
import Debug from "debug";
import "mocha";
import {SocketClient} from "../src/SocketClient";
import {TestHelper} from "./Helpers/TestHelper";

const debug = Debug("app:Test");

// prepare variables
const host = (process.env.HOST || "").trim();
const socketClient0a: SocketClient = new SocketClient(host, "abc");
const socketClient1a: SocketClient = new SocketClient(host, "1");
const socketClient1b: SocketClient = new SocketClient(host, "1");
const socketClient2a: SocketClient = new SocketClient(host, "2");
const socketClient3a: SocketClient = new SocketClient(host, "3");

describe("Socket Client", () => {
  beforeEach(async () => {
    // runs before all tests in this block
  });

  afterEach(async () => {
    // runs after all tests in this block
    socketClient1a.disconnect();
    socketClient1b.disconnect();
    socketClient2a.disconnect();
    socketClient3a.disconnect();
    debug("----------------");
    await TestHelper.timeout(100);
  });

  it("invalid connection", async () => {
    await TestHelper.AssertInvalidConnect(socketClient0a);
  }).timeout(30 * 1000); // the first connection make take longer since the server may need cold start

  it("connect fight", async () => {
    // socketClient1a
    await TestHelper.AssertConnect(socketClient1a);
    const promise = new Promise(resolve => socketClient1a.on("disconnect", resolve));
    await TestHelper.AssertConnect(socketClient1b);

    // socketClient1a
    await promise;
    assert.isFalse(socketClient1a.isConnected);
  });

  it("resume room", async () => {
    // connect
    await TestHelper.AssertConnect(socketClient1a);
    await TestHelper.AssertLeaveIfHasRoom(socketClient1a);

    // create room
    await TestHelper.AssertCreateRoom(socketClient1a);

    // disconnect
    await TestHelper.AssertDisconnect(socketClient1a);

    // connect and check has room
    await TestHelper.AssertConnect(socketClient1a);
    await TestHelper.AssertHasRoom(socketClient1a);
    await TestHelper.AssertLeaveRoom(socketClient1a);
  });

  it("join room and chat", async () => {
    // socketClient1a
    await TestHelper.AssertConnect(socketClient1a);
    await TestHelper.AssertLeaveIfHasRoom(socketClient1a);
    await TestHelper.AssertCreateRoom(socketClient1a);

    // socketClient2a
    await TestHelper.AssertConnect(socketClient2a);
    await TestHelper.AssertLeaveIfHasRoom(socketClient2a);
    await TestHelper.AssertJoinRoom(socketClient2a, socketClient1a.roomId);

    // socketClient3a
    await TestHelper.AssertConnect(socketClient3a);
    await TestHelper.AssertInvalidJoinRoom(socketClient3a, socketClient1a.roomId);

    const promise1 = new Promise(resolve => socketClient1a.once("respondMessageInRoom", resolve));
    const promise2 = new Promise(resolve => socketClient2a.once("respondMessageInRoom", resolve));

    // socketClient: send chat
    const newMessage = "Hello World!";
    await TestHelper.AssertChatInRoom(socketClient2a, newMessage);
    await Promise.all([promise1, promise2]);
    assert.isAtLeast(socketClient1a.chats.length, 1);
    assert.isAtLeast(socketClient2a.chats.length, 1);
    assert.equal(socketClient1a.chats[0], newMessage);
  });
});

File: 2019_07_SocketIO/James/Tests/TestConnection.cs
using System;
using System.Collections;
using NUnit.Framework;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.TestTools;

namespace Tests
{
    [Serializable]
    public class SearchResult
    {
        public string name;
        public string latest;
    }

    public struct SearchJson
    {
        public SearchResult[] results;
        public int total;
    }
    
    public class TestConnection
    {
        private const string Host = "https://blooming-fjord-24277.herokuapp.com/socket.io/";

        public GameClient gameClient0a = new GameClient(Host, "abc");
        public GameClient gameClient1a = new GameClient(Host, "1");
        public GameClient gameClient1b = new GameClient(Host, "1");
        public GameClient gameClient2a = new GameClient(Host, "2");
        public GameClient gameClient3a = new GameClient(Host, "3");

        [SetUp]
        public void SetUp() {
            //SetUp runs before all test cases
            Debug.Log($"--- Setup --- ");
        }

        [TearDown]
        public void TearDown() {
            //SetUp runs after all test cases
            Debug.Log($"--- TearDown --- ");
            
            gameClient1a.Disconnect();
            gameClient1b.Disconnect();
            gameClient2a.Disconnect();
            gameClient3a.Disconnect();
        }
        
        [UnityTest]
        public IEnumerator TestSimpleJson()
        {
            string url = "https://api.cdnjs.com/libraries?search=mocha";
            UnityWebRequest www = UnityWebRequest.Get(url);
            yield return www.SendWebRequest();

            if (www.isNetworkError || www.isHttpError)
            {
                Debug.Log(www.error);
                Assert.IsTrue(false);
            }
            else
            {
                yield return TestHelper.Timeout(1.0);
                
                // Show results as text
                var json = JsonUtility.FromJson<SearchJson>(www.downloadHandler.text);
                Debug.Log($"{www.downloadHandler.text}");
                Assert.GreaterOrEqual(json.total, 1);
                Assert.GreaterOrEqual(json.results.Length, 1);
                Assert.AreEqual(json.results[0].name, "mocha");
            }

            // Use the Assert class to test conditions.
            // yield to skip a frame
            yield return null;
        }

        [UnityTest]
        public IEnumerator TestInvalidConnection()
        {
            yield return GameClientHelper.AssertInvalidConnect(gameClient0a);
        }

        [UnityTest]
        public IEnumerator TestConnectFight()
        {
            yield return GameClientHelper.AssertConnect(gameClient1a);
            yield return GameClientHelper.AssertConnect(gameClient1b);

            yield return TestHelper.Timeout(() => !gameClient1a.IsConnected);
            Assert.IsFalse(gameClient1a.IsConnected);
        }

        [UnityTest]
        public IEnumerator TestResumeRoom()
        {
            yield return GameClientHelper.AssertConnect(gameClient1a);
            yield return GameClientHelper.AssertLeaveIfHasRoom(gameClient1a);
            yield return GameClientHelper.AssertCreateRoom(gameClient1a);
            
            // disconnect
            yield return GameClientHelper.AssertDisconnect(gameClient1a);
            
            // connect and check has room
            yield return GameClientHelper.AssertConnect(gameClient1a);
            yield return GameClientHelper.AssertHasRoom(gameClient1a);
            yield return GameClientHelper.AssertLeaveRoom(gameClient1a);
        }

        [UnityTest]
        public IEnumerator TestJoinRoomAndChat()
        {
            // socketClient1a
            yield return GameClientHelper.AssertConnect(gameClient1a);
            yield return GameClientHelper.AssertLeaveIfHasRoom(gameClient1a);
            yield return GameClientHelper.AssertCreateRoom(gameClient1a);
            
            // socketClient2a
            yield return GameClientHelper.AssertConnect(gameClient2a);
            yield return GameClientHelper.AssertLeaveIfHasRoom(gameClient2a);
            yield return GameClientHelper.AssertJoinRoom(gameClient2a, gameClient1a.RoomId);
            
            // socketClient3a
            yield return GameClientHelper.AssertConnect(gameClient3a);
            yield return GameClientHelper.AssertInvalidJoinRoom(gameClient3a, gameClient1a.RoomId);
            
            // socketClient: send chat
            int total = 0;
            gameClient1a.OnEventMessage += (sender, message) =>
            {
                if (message.EventName == "respondMessageInRoom")
                    total++;
            };
            gameClient2a.OnEventMessage += (sender, message) =>
            {
                if (message.EventName == "respondMessageInRoom")
                    total++;
            };
            
            string newMessage = "Hello World!";
            yield return GameClientHelper.AssertChatInRoom(gameClient1a, newMessage);
            yield return TestHelper.Timeout(() => total == 2);

            Assert.AreEqual(gameClient1a.Chats.Count, 1);
            Assert.AreEqual(gameClient2a.Chats.Count, 1);
            Assert.AreEqual(gameClient1a.Chats[0], newMessage);
        }
    }
}